improve spotify support

fugkco opened the pull request:
Hello, this PR improves the support for Spotify. Support for individual tracks and albums has been added. I've also removed the reliance on BS4, and instead it uses Spotify's API, which means the code will be less likely to break.

EDIT: forgot to mention, this addresses #55

kmille commented:
Hey,
can you give me a link to a spotify playlist to reproduce the issue and test the PR? I don't use Spotify at all.

fugkco commented:
This is not the same playlist, but I quickly put this one together, it has a roughly similar number of songs in the playlist, and should produce the same thing:

```
spotify:playlist:1u1G4qggyzs2q6wDm5R0YT
```

kmille commented on app/spotify.py:
```diff
@@ -22,34 +65,65 @@ def get_songs_from_spotify_website(playlist):
+            if resp['next'] is None:
+                break
+            url = resp['next']
+    elif url_info["type"] == "track":
```
Why do you make only a single retry for tracks/albums but infinity retries for a playlist download?

fugkco replied:
The single retry is because of rate limiting. If we wait long enough (or specifically the number of seconds returned in the `Retry-After` is enough), then we should be able to receive the correct response. For playlists it's about retry-after header but more so it's about the pagination. In theory a playlist that has less than 100 items will be retried only once for the rate limiting just like tracks/albums.

kmille replied:
ah okay

kmille commented on app/spotify.py:
```diff
@@ -22,34 +65,65 @@ def get_songs_from_spotify_website(playlist):
+    req = requests.get(api_url, headers={'Authorization': 'Bearer {}'.format(access_token)})
+
+    if req.status_code == 429:
+        sleep(int(req.headers.get("Retry-After")) + 1)
```
I don't know how much rate limiting is a thing for the spotify api. If we got blocked regularly, maybe it's better to add a print statement here.

fugkco replied:
During my development here I didn't get rate limited once. I haven't really been able to fully test it. Maybe a print statement is a good idea regardless. Will add.

kmille commented on app/tests.py:
```diff
@@ -255,12 +282,12 @@ def test_spotify_parser_valid_playlist_id(self):
     def test_spotify_parser_invalid_playlist_id(self):
         playlist_id = "thisdoesnotexist"
-        with self.assertRaises(SpotifyWebsiteParserException):
+        with self.assertRaises(SpotifyInvalidUrlException):
```
Does this work for you? In my case a SpotifyWebsiteParserException is raised like before

fugkco replied:
Oh, it's probably cos I added backwards compatibility later on. Strange though, I could have sworn all tests passed prior to submitting. Will fix.

kmille reviewed (commented):
I get the following error if I download the playlist (as zip):
```bash
127.0.0.1 - - [02/Apr/2021 20:25:58] "POST /playlist/spotify HTTP/1.1" 200 -                                                                                             
Worker 0 is now working on task: {'playlist_name': 'test playlist', 'playlist_id': 'spotify:playlist:1u1G4qggyzs2q6wDm5R0YT', 'add_to_playlist': False, 'create_zip': True}                                                                                                                                                                       
127.0.0.1 - - [02/Apr/2021 20:26:00] "GET /debug HTTP/1.1" 200 -                                                                                                         
127.0.0.1 - - [02/Apr/2021 20:26:01] "GET /debug HTTP/1.1" 200 -                                                                                                         
We got 1482 songs from the Spotify playlist                                                                                                                              
Downloading 'Dua Lipa - Break My Heart.mp3'                                                                                                                              
Dowload finished: /tmp/deezer-downloader/playlists/test playlist/Dua Lipa - Break My Heart.mp3                                                                           
Downloading 'Justin Bieber - Anyone.mp3'
...
Downloading 'Olivia Rodrigo - deja vu.mp3'
Got an Exception in function download_spotify_playlist_and_queue_and_zip with parameters '{'playlist_name': 'test playlist', 'playlist_id': 'spotify:playlist:1u1G4qggyzs2q6wDm5R0YT', 'add_to_playlist': False, 'create_zip': True}
'NoneType' object is not subscriptable'
worker 0 is done with task: {'playlist_name': 'test playlist', 'playlist_id': 'spotify:playlist:1u1G4qggyzs2q6wDm5R0YT', 'add_to_playlist': False, 'create_zip': True} (state=failed)
Worker 0 is waiting for a task```

fugkco commented:
> I get the following error if I download the playlist (as zip):

Will look into that!

kmille commented:
I don't know what's the problem. It works just fine after a second  try.

fugkco commented:
Yeah, I can't reproduce it either. I'm not sure what might have caused it, but I have a feeling that for _some_ reason one of the entries was `None`, causing that. I've added a safeguard in case that's the issue. By the way, I've removed beautiful soup since it was only in the Spotify code. I couldn't find any other references. Let me know if I've missed something and if it's still needed.

kmille commented:
Thank You for the contribution!
